RecoTracker/CkfPattern/src/CkfTrackCandidateMakerBase.cc

```diff
@@ -540,4 +540,32 @@ namespace cms {
     e.put(std::move(outputSeedStopInfos));
   }
 
+  void CkfTrackCandidateMakerBase::fillPSetDescription(edm::ParameterSetDescription& desc) {
+    desc.add<bool>("cleanTrajectoryAfterInOut", true);
+    desc.add<bool>("doSeedingRegionRebuilding", true);
+    desc.add<bool>("onlyPixelHitsForSeedCleaner", false);
+    desc.add<bool>("reverseTrajectories", false);
+    desc.add<bool>("useHitsSplitting", true);
+    desc.add<edm::InputTag>("MeasurementTrackerEvent", edm::InputTag("MeasurementTrackerEvent"));
+    desc.add<edm::InputTag>("clustersToSkip", edm::InputTag(""));
+    desc.add<edm::InputTag>("phase2clustersToSkip", edm::InputTag(""));
+    desc.add<edm::InputTag>("src", edm::InputTag("globalMixedSeeds"));
+
+    edm::ParameterSetDescription psd0;
+    desc.add<edm::ParameterSetDescription>("TrajectoryBuilderPSet", psd0);
```

Comment:
I'm afraid this states that the `TrajectoryBuilderPSet` can be only an empty PSet.

In the `CkfTrackCandidates_cfi` the `TrajectoryBuilderPSet = cms.PSet(refToPSet_ = cms.string('GroupedCkfTrajectoryBuilder'))` means that the contents of `TrajectoryBuilderPSet` are taken from `process.GroupedCkfTrajectoryBuilder` PSet at the time the python configuration is "serialized" to be shipped to the C++ code, where the configuration validation, based on the `ConfigurationDescriptions`, is run. The configuration validation should then fail because the PSet has unexpected parameters.

A complication here is that the parameters of this PSet depend (in general, at least, I didn't check this particular case) on the helper plugin. But, in this case the helper plugin type appears to be defined within the PSet, in which case it might be straightforward to use 
https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He

Reply:
@makortel 

> The configuration validation should then fail because the PSet has unexpected parameters.

indeed it fails at runtime with:
```console
----- Begin Fatal Exception 23-Sep-2021 18:17:08 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=CkfTrackCandidateMaker label='convTrackCandidates'
Exception Message:
Illegal parameters found in configuration.  The parameters are named:
 'ComponentType'
 'MeasurementTrackerName'
 'TTRHBuilder'
 'alwaysUseInvalidHits'
 'bestHitOnly'
 'estimator'
 'foundHitBonus'
 'inOutTrajectoryFilter'
 'intermediateCleaning'
 'keepOriginalIfRebuildFails'
 'lockHits'
 'lostHitPenalty'
 'maxCand'
 'minNrOfHitsForRebuild'
 'propagatorAlong'
 'propagatorOpposite'
 'requireSeedHitsInRebuild'
 'seedAs5DHit'
 'trajectoryFilter'
 'updator'
 'useSameTrajFilter'
You could be trying to use parameter names that are not
allowed for this plugin or they could be misspelled.
----- End Fatal Exception -------------------------------------------------
```

> I'm afraid this states that the TrajectoryBuilderPSet can be only an empty PSet.

now, this is quite unfortunate because in turn `process.GroupedCkfTrajectoryBuilder` depends on other `refToPSet_`s, e.g.:

https://github.com/cms-sw/cmssw/blob/1a6b75da950e434d4b5dbc1d6220ce80caa51d60/RecoTracker/CkfPattern/python/GroupedCkfTrajectoryBuilder_cfi.py#L12

and

https://github.com/cms-sw/cmssw/blob/1a6b75da950e434d4b5dbc1d6220ce80caa51d60/RecoTracker/CkfPattern/python/GroupedCkfTrajectoryBuilder_cfi.py#L14

> . But, in this case the helper plugin type appears to be defined within the PSet, in which case it might be straightforward to use https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideConfigurationValidationAndHelp#A_Plugin_Module_Using_Another_He

Let me get this straight, where should I define the descriptions for the configurations that are just contained in a `PSet` which doesn't seem to be directly linked with any `EDModule` ?

Reply:
> Let me get this straight, where should I define the descriptions for the configurations that are just contained in a `PSet` which doesn't seem to be directly linked with any `EDModule` ?

Strictly speaking there is no mechanism to define the `fillDescriptions()` for such "global PSets". But such global PSets can still be used via `cms.PSet(refToPSet_ = cms.string(...))` construct in an EDModule that uses the "helper plugin mechanism" in its `fillDescriptions()`.

In the "helper plugin mechanism" the `static void fillPSetDescription(edm::ParameterSetDescription& iDesc)` would be defined in all the plugin classes that are part of `BaseCkfTrajectoryBuilderFactory` factory, the factory registration macro
https://github.com/cms-sw/cmssw/blob/1a6b75da950e434d4b5dbc1d6220ce80caa51d60/RecoTracker/CkfPattern/src/BaseCkfTrajectoryBuilderFactory.cc#L3
needs to be changed to `EDM_REGISTER_VALIDATED_PLUGINFACTORY`, and the plugin registration macro e.g. in
https://github.com/cms-sw/cmssw/blob/1a6b75da950e434d4b5dbc1d6220ce80caa51d60/RecoTracker/CkfPattern/plugins/SealModules.cc#L21-L22
need to be changed to `DEFINE_EDM_VALIDATED_PLUGIN`.

When using a global PSet via `refToPSet_` in this setup the PSet is missing a parameter, that parameter is inserted into "a local copy of the PSet" during the configuration validation phase.

But a new parameter does not get added into the "global PSet" in the cfi generation phase. A new parameter gets added only to the cfi generated for the EDModule.

Reply:
I see, thanks. I will take a stab at it tomorrow.

Reply:
@mmusich
In case it helps, I started to follow @makortel's instructions in 435217848c0.
It does not work yet, but maybe it saves you some typing.

Reply:
Thanks, @missirol I merged your commit, though I am a bit lost though on what still needs to be done.
If I merge your changes on top of mine and I run for example wf. 4.53 I get:
```
 ----- Begin Fatal Exception 24-Sep-2021 17:15:28 CEST-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=CkfTrackCandidateMaker label='convTrackCandidates'
Exception Message:
Illegal parameter found in configuration.  The parameter is named:
 'minGoodStripCharge'
You could be trying to use a parameter name that is not
allowed for this plugin or it could be misspelled.
----- End Fatal Exception -------------------------------------------------
```
while I would have naively expected that to be taken care from here:

https://github.com/cms-sw/cmssw/commit/435217848c0d3551436d02f59d49cf121f59b798#diff-3a57ce7e190928d1e542d59a175b5be0b6b5d0c6bb0a664aa909e96778f92c6eR42

Reply:
When I last tested, I encountered a similar complaint (in a different wf, from a different module, about different parameters), so something must be off in how I filled the `FillDescriptions` methods. I'll try to have a look today/tomorrow.